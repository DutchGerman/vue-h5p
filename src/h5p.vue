<template>
  <div v-if="loading">
    <slot/>
  </div>
  <div v-else-if="getJSONErrorResponse">
    <slot name="404" :response="getJSONErrorResponse"/>
  </div>
  <iframe v-else ref="iframe" style="width: 100%; height: 100%; border: none;" :srcdoc="srcdoc" @load="addEventHandlers"/>
</template>

<script>
import l10n from './l10n'
// eslint-disable-next-line
import frameScript from '!raw-loader!../frame/frame'
// eslint-disable-next-line
import frameStyle from '!to-string-loader!css-loader!../vendor/h5p/styles/h5p.css'
import Toposort from 'toposort-class'

export default {
  name: 'h5p',
  props: {
    src: {
      type: String,
      required: true
    },
    displayOptions: {
      type: Object,
      default: () => {
        return {
          frame: true,
          export: true,
          embed: true,
          copyright: true,
          icon: true
        }
      }
    },
    l10n: {
      type: Object,
      default: () => { return {} }
    },
    contentId: {
      type: [String, Number],
      default: 1
    }
  },
  data () {
    return {
      mainLibrary: undefined,
      h5pIntegration: { l10n: { H5P: {} } },
      started: false,
      loading: true,
      getJSONErrorResponse: undefined,
      srcdoc: ''
    }
  },
  computed: {
    path () {
      return this.src.slice(this.src.length - 1) === '/' ? this.src.substring(0, this.src.length - 1) : this.src
    },
    contentStyles () {
      return this.h5pIntegration.contents['cid-' + this.contentId].styles
    },
    contentScripts () {
      return this.h5pIntegration.contents['cid-' + this.contentId].scripts
    },
    head () {
      // workaround for vue-loader parsing this as the end of our SFC's script block
      const endScript = '</' + 'script>'

      const styleTags = [
        `<style>${frameStyle}</style>`,
        ...this.contentStyles.map(style => {
          return `<link rel="stylesheet" href="${style}">`
        })
      ]

      const scriptTags = [
        `<script>H5PIntegration = ${JSON.stringify(this.h5pIntegration)};${endScript}`,
        `<script>${frameScript}${endScript}`,
        ...this.contentScripts.map(script => {
          return `<script src="${script}">${endScript}`
        })
      ]

      return [
        '<head>',
        '<base target="_parent">',
        ...styleTags,
        ...scriptTags,
        '</head>'
      ]
    }
  },
  async mounted () {
    let content
    let dependencies
    try {
      this.h5p = await this.getJSON(`${this.path}/h5p.json`)
      content = await this.getJSON(`${this.path}/content/content.json`)
      this.h5pIntegration.pathIncludesVersion = this.pathIncludesVersion = await this.checkIfPathIncludesVersion()
      this.mainLibrary = await this.findMainLibrary()
      dependencies = await this.findAllDependencies()
    } catch (e) {
      if (e.message === '404') {
        this.loading = false
        return
      } else {
        throw e
      }
    }

    const { styles, scripts } = this.sortDependencies(dependencies)

    this.h5pIntegration.url = this.path
    this.h5pIntegration.contents = this.h5pIntegration.contents || {}

    this.h5pIntegration.contents[`cid-${this.contentId}`] = {
      library: `${this.mainLibrary.machineName} ${this.mainLibrary.majorVersion}.${this.mainLibrary.minorVersion}`,
      jsonContent: JSON.stringify(content),
      styles: styles,
      scripts: scripts,
      displayOptions: this.displayOptions
    }

    Object.assign(this.h5pIntegration.l10n.H5P, l10n.H5P, this.l10n)

    this.srcdoc = [
      '<!doctype html><html class="h5p-iframe">',
      ...this.head,
      `<body><div class="h5p-content" data-content-id="${this.contentId}"/></body></html>`
    ].join('')

    this.loading = false
  },
  methods: {
    addEventHandlers () {
      this.$refs.iframe.contentWindow.H5P.externalDispatcher.on('*', (ev) => {
        this.$emit(ev.type.toLowerCase(), ev.data)
      })
    },
    async getJSON (url) {
      const resp = await fetch(url, { credentials: 'include' })
      if (resp.status === 404) {
        this.getJSONErrorResponse = resp
        throw new Error(404)
      }
      const json = await resp.json()
      return json
    },
    async checkIfPathIncludesVersion () {
      const dependency = this.h5p.preloadedDependencies[0]
      const machinePath = dependency.machineName + '-' + dependency.majorVersion + '.' + dependency.minorVersion

      let pathIncludesVersion

      try {
        await fetch(`${this.path}/${machinePath}/library.json`, { credentials: 'include', method: 'head' })
        pathIncludesVersion = true
      } catch (e) {
        pathIncludesVersion = false
      }
      return pathIncludesVersion
    },
    findMainLibrary () {
      const mainLibraryInfo = this.h5p.preloadedDependencies.find(dep => dep.machineName === this.h5p.mainLibrary)

      this.mainLibraryPath = this.h5p.mainLibrary + (this.pathIncludesVersion ? '-' + mainLibraryInfo.majorVersion + '.' + mainLibraryInfo.minorVersion : '')
      return this.getJSON(`${this.path}/${this.mainLibraryPath}/library.json`)
    },
    findAllDependencies () {
      const directDependencyNames = this.h5p.preloadedDependencies.map(dependency => this.libraryPath(dependency))

      return this.loadDependencies(directDependencyNames, [])
    },
    libraryPath (library) {
      return library.machineName + (this.pathIncludesVersion ? '-' + library.majorVersion + '.' + library.minorVersion : '')
    },
    async loadDependencies (toFind, alreadyFound) {
      const dependencies = alreadyFound
      const findNext = []
      const newDependencies = await Promise.all(toFind.map((libraryName) => this.findLibraryDependencies(libraryName)))
      // loop over newly found libraries
      newDependencies.forEach((library) => {
        // push into found list
        dependencies.push(library)
        // check if any dependencies haven't been found yet
        library.dependencies.forEach((dependency) => {
          if (!dependencies.find((foundLibrary) => foundLibrary.libraryPath === dependency) && !newDependencies.find((foundLibrary) => foundLibrary.libraryPath === dependency)) {
            findNext.push(dependency)
          }
        })
      })

      if (findNext.length > 0) {
        return this.loadDependencies(findNext, dependencies)
      }
      return dependencies
    },
    async findLibraryDependencies (libraryName) {
      const library = await this.getJSON(`${this.path}/${libraryName}/library.json`)
      const libraryPath = this.libraryPath(library)

      let dependencies = []
      if (library.preloadedDependencies) {
        dependencies = library.preloadedDependencies.map(dependency => this.libraryPath(dependency))
      }

      return { libraryPath, dependencies, preloadedCss: library.preloadedCss, preloadedJs: library.preloadedJs }
    },
    sortDependencies (dependencies) {
      const dependencySorter = new Toposort()
      const CSSDependencies = {}
      const JSDependencies = {}

      dependencies.forEach(dependency => {
        dependencySorter.add(dependency.libraryPath, dependency.dependencies)

        if (dependency.preloadedCss) {
          CSSDependencies[dependency.libraryPath] = CSSDependencies[dependency.libraryPath] ? CSSDependencies[dependency.libraryPath] : []
          dependency.preloadedCss.forEach(style => {
            CSSDependencies[dependency.libraryPath].push(`${this.path}/${dependency.libraryPath}/${style.path}`)
          })
        }

        if (dependency.preloadedJs) {
          JSDependencies[dependency.libraryPath] = JSDependencies[dependency.libraryPath] ? JSDependencies[dependency.libraryPath] : []
          dependency.preloadedJs.forEach(script => {
            JSDependencies[dependency.libraryPath].push(`${this.path}/${dependency.libraryPath}/${script.path}`)
          })
        }
      })

      const styles = []
      const scripts = []

      dependencySorter.sort().reverse().forEach(function (dependencyName) {
        Array.prototype.push.apply(styles, CSSDependencies[dependencyName])
        Array.prototype.push.apply(scripts, JSDependencies[dependencyName])
      })

      if (this.mainLibrary.preloadedCss) {
        Array.prototype.push.apply(styles, this.mainLibrary.preloadedCss.map(style => `${this.path}/${this.mainLibraryPath}/${style.path}`))
      }
      if (this.mainLibrary.preloadedJs) {
        Array.prototype.push.apply(scripts, this.mainLibrary.preloadedJs.map(script => `${this.path}/${this.mainLibraryPath}/${script.path}`))
      }
      return { styles, scripts }
    }
  }
}
</script>
