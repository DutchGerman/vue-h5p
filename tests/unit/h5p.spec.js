import { shallowMount } from '@vue/test-utils'
import h5p from '@/h5p'
import flushPromises from 'flush-promises'

jest.mock('!to-string-loader!css-loader!../vendor/h5p/styles/h5p.css', () => {
  return 'MOCKED_FRAME_CSS'
}, {
  virtual: true
})

jest.mock('!raw-loader!../frame/frame', () => {
  return 'MOCKED_FRAME_JS'
}, {
  virtual: true
})

function createComponent (props, cbDefault, cbError) {
  const renderDefault = jest.fn(cbDefault)
  const renderError = jest.fn(cbError)
  const wrapper = shallowMount(h5p, {
    propsData: props,
    scopedSlots: {
      default: renderDefault,
      error: renderError
    }
  })
  return {
    renderDefault,
    renderError,
    wrapper
  }
}

describe('Component', () => {
  let wrapper
  let renderDefault
  let renderError

  beforeEach(() => {
    // just reset .mock data, but not .mockResponse
    fetch.mockClear()
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('renders an iframe for existing h5p-content', async () => {
    ({ wrapper } = createComponent({
      src: 'http://test.example/h5p'
    }))
    await flushPromises()
    expect(wrapper.element.constructor.name).toBe('HTMLIFrameElement')
  })

  it('renders default slot content while loading', async () => {
    ({ renderDefault, wrapper } = createComponent({
      src: 'http://test.example/h5p'
    }))
    expect(wrapper.vm.loading).toBe(true)
    expect(renderDefault).toHaveBeenCalled()
    await flushPromises()
    expect(wrapper.vm.loading).toBe(false)
    expect(renderDefault).toHaveBeenCalledTimes(1)
  })

  it('renders error slot on fetch-errors and provides response object', async () => {
    ({ renderError, wrapper } = createComponent({
      src: 'http://test.example/404'
    }))
    await flushPromises()
    expect(wrapper.vm.loading).toBe(false)
    expect(renderError).toHaveBeenCalledWith(expect.objectContaining({
      response: expect.any(Object)
    }))
  })

  it('does not render error slot on succesful fetch', async () => {
    ({ renderError, wrapper } = createComponent({
      src: 'http://test.example/h5p'
    }))
    await flushPromises()
    expect(wrapper.vm.loading).toBe(false)
    expect(renderError).toHaveBeenCalledTimes(0)
  })

  it('correctly handles the src prop', async () => {
    ({ wrapper } = createComponent({
      src: 'http://test.example/h5p'
    }))
    await flushPromises()
    expect(fetch).toHaveBeenCalledWith('http://test.example/h5p/h5p.json', expect.anything())
    wrapper.destroy()
    fetch.mockClear();

    ({ wrapper } = createComponent({
      src: 'http://test.example/h5p/'
    }))
    await flushPromises()
    expect(fetch).toHaveBeenCalledWith('http://test.example/h5p/h5p.json', expect.anything())
  })

  it('requests the neccessary json files', async () => {
    ({ wrapper } = createComponent({
      src: 'http://test.example/h5p'
    }))
    await flushPromises()
    expect(fetch).toHaveBeenCalledWith('http://test.example/h5p/h5p.json', expect.anything())
    expect(fetch).toHaveBeenCalledWith('http://test.example/h5p/content/content.json', expect.anything())
    expect(fetch).toHaveBeenCalledWith('http://test.example/h5p/H5P.Image-1.1/library.json', expect.anything())
  })

  describe('iframe', () => {
    it('has correct attribute srcdoc', async () => {
      ({ wrapper } = createComponent({
        src: 'http://test.example/h5p'
      }))
      await flushPromises()
      expect(wrapper.element.srcdoc).toMatchSnapshot()
    })

    describe('window has correct h5pIntegration object', () => {
      // TODO: how to wait for iframe load event to get access to iframe documentWindow?
    })
  })
})
