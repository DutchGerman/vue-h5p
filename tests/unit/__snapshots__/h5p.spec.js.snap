// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component iframe has correct attribute srcdoc 1`] = `
<!doctype html>
<html class="h5p-iframe">

  <head>
    <base target="_parent">
    <style>
      MOCKED_FRAME_CSS
    </style>
    <link rel="stylesheet" href="http://test.example/h5p/H5P.Image-1.1/image.css">
    <link rel="stylesheet" href="http://test.example/h5p/H5P.Image-1.1/image.css">
    <script>
      H5PIntegration = {
        "l10n": {
          "H5P": {
            "advancedHelp": "Include this script on your website if you want dynamic sizing of the embedded content:",
            "author": "Author",
            "by": "by",
            "cancelLabel": "Cancel",
            "changes": "Changelog",
            "close": "Close",
            "confirmDialogBody": "Please confirm that you wish to proceed. This action is not reversible.",
            "confirmDialogHeader": "Confirm action",
            "confirmLabel": "Confirm",
            "connectionLost": "Connection lost. Results will be stored and sent when you regain connection.",
            "connectionReestablished": "Connection reestablished.",
            "contentChanged": "This content has changed since you last used it.",
            "contentCopied": "Content is copied to the clipboard",
            "contentType": "Content Type",
            "copyrightInformation": "Rights of use",
            "copyrights": "Rights of use",
            "copyrightsDescription": "View copyright information for this content.",
            "disableFullscreen": "Disable fullscreen",
            "download": "Download",
            "downloadDescription": "Download this content as a H5P file.",
            "embed": "Embed",
            "embedDescription": "View the embed code for this content.",
            "fullscreen": "Fullscreen",
            "h5pDescription": "Visit H5P.org to check out more cool content.",
            "hideAdvanced": "Hide advanced",
            "license": "License",
            "licenseC": "Copyright",
            "licenseCC010": "CC0 1.0 Universal (CC0 1.0) Public Domain Dedication",
            "licenseCC10": "1.0 Generic",
            "licenseCC20": "2.0 Generic",
            "licenseCC25": "2.5 Generic",
            "licenseCC30": "3.0 Unported",
            "licenseCC40": "4.0 International",
            "licenseCCBY": "Attribution",
            "licenseCCBYNC": "Attribution-NonCommercial",
            "licenseCCBYNCND": "Attribution-NonCommercial-NoDerivs",
            "licenseCCBYNCSA": "Attribution-NonCommercial-ShareAlike",
            "licenseCCBYND": "Attribution-NoDerivs",
            "licenseCCBYSA": "Attribution-ShareAlike",
            "licenseExtras": "License Extras",
            "licenseGPL": "General Public License",
            "licensePD": "Public Domain",
            "licensePDM": "Public Domain Mark",
            "licenseU": "Undisclosed",
            "licenseV1": "Version 1",
            "licenseV2": "Version 2",
            "licenseV3": "Version 3",
            "noCopyrights": "No copyright information available for this content.",
            "offlineDialogBody": "We were unable to send information about your completion of this task. Please check your internet connection.",
            "offlineDialogHeader": "Your connection to the server was lost",
            "offlineDialogRetryButtonLabel": "Retry now",
            "offlineDialogRetryMessage": "Retrying in :num....",
            "offlineSuccessfulSubmit": "Successfully submitted results.",
            "resubmitScores": "Attempting to submit stored results.",
            "reuse": "Reuse",
            "reuseContent": "Reuse Content",
            "reuseDescription": "Reuse this content.",
            "showAdvanced": "Show advanced",
            "showLess": "Show less",
            "showMore": "Show more",
            "size": "Size",
            "source": "Source",
            "startingOver": "You'll be starting over.",
            "subLevel": "Sublevel",
            "thumbnail": "Thumbnail",
            "title": "Title",
            "year": "Year"
          }
        },
        "pathIncludesVersion": true,
        "url": "http://test.example/h5p",
        "contents": {
          "cid-default": {
            "library": "H5P.Image 1.1",
            "jsonContent": "{\\"presentation\\":{\\"slides\\":[]}}",
            "styles": ["http://test.example/h5p/H5P.Image-1.1/image.css", "http://test.example/h5p/H5P.Image-1.1/image.css"],
            "scripts": ["http://test.example/h5p/H5P.Image-1.1/image.js", "http://test.example/h5p/H5P.Image-1.1/image.js"],
            "displayOptions": {}
          }
        }
      };
    </script>
    <script>
      MOCKED_FRAME_JS
    </script>
    <script src="http://test.example/h5p/H5P.Image-1.1/image.js"></script>
    <script src="http://test.example/h5p/H5P.Image-1.1/image.js"></script>
  </head>

  <body>
    <div class="h5p-content" data-content-id="default" />
  </body>

</html>
`;
